Std/Data/BitVec/Bitblast.lean

```diff
@@ -0,0 +1,122 @@
+  cases b <;> simp
+
+@[simp] theorem toNat_eq_bif {b: Bool} : cond b 1 0 = toNat b := by simp [cond, toNat]
+
+theorem shiftRight_eq_testBit (x i : Nat) : (x >>> i) % 2 = toNat (testBit x i) := by
+  simp [Nat.testBit, Nat.mod_two_of_bodd, toNat_eq_bif]
+
+theorem div_add_mod_two_pow (m n : Nat) : n = 2 ^ m * n >>> m + n % (2 ^ m):= by
+  simp [shiftRight_eq_div_pow, Nat.div_add_mod]
+
+theorem mod_two_pow_succ (x i : Nat) :
+  x % 2 ^ (i + 1) = 2 ^ i * toNat (Nat.testBit x i) + x % (2 ^ i):= by
+  have h1 := div_add_mod_two_pow i x
+  rw [←div_add_mod (x >>> i) 2, Nat.mul_add, ←Nat.mul_assoc, ←pow_succ, shiftRight_eq_testBit] at h1
+  have h2 := Eq.trans (div_add_mod_two_pow (i + 1) x).symm h1
+  rw [shiftRight_succ, succ_eq_add_one, Nat.add_assoc] at h2
+  exact Nat.add_left_cancel h2
+
+/-- Generic method to create a natural number by appending bits tail-recursively.
+It takes a boolean function `f` on each bit the number of bits `i`.  It is
+almost always specialized  `i = w`; the length of the binary representation.
+This is an alternative to using `List` and is used to define bitadd, bitneg, bitmul etc.-/
+def ofBits (f : Nat → Bool) (z : Nat) : Nat → Nat
+  | 0 => z
+  | i + 1 => ofBits f (z.bit (f i)) i
+
+theorem ofBits_succ {f : Nat → Bool} {i z : Nat} : ofBits f z i = 2 ^ i * z + ofBits f 0 i:= by
+  revert z
+  induction i with
+  | zero => simp [ofBits, bit_val]
+  | succ i ih =>  intro z; simp only [ofBits, @ih (bit (f i) 0), @ih (bit (f i) z)]
+                  rw [bit_val, Nat.mul_add, ← Nat.mul_assoc, ← pow_succ]
+                  simp [bit_val, Nat.add_assoc]
+
+theorem ofBits_lt {f : Nat → Bool} : ofBits f 0 i < 2^i := by
+  induction i with
+  | zero => simp [ofBits, bit_val, lt_succ]
+  | succ i ih =>  simp only [ofBits]
+                  rw [ofBits_succ, two_pow_succ]
+                  cases (f i)
+                  · exact add_lt_add (by simp [bit_val, two_pow_pos i]) ih
+                  · simp only [bit_val, Nat.mul_zero, cond_true, Nat.zero_add, Nat.mul_one]
+                    exact Nat.add_lt_add_left ih _
+
+theorem ofBits_testBit {f : Nat → Bool} (h1: i < j) : (ofBits f 0 j).testBit i = f i := by
+  cases j with
+  | zero => simp [not_lt_zero] at h1
+  | succ j =>
+    revert i
+    induction j with
+    | zero => simp [lt_succ, ofBits, bit_val]
+              cases (f 0) <;> trivial
+    | succ j ih =>
+      intro i hi
+      rw [lt_succ] at hi
+      cases Nat.lt_or_eq_of_le hi with
+      | inl h1 => rw [← ih h1, ofBits, ofBits_succ, testBit_two_pow_mul_add h1]
+      | inr h1 => rw [h1, ofBits, ofBits_succ, testBit_two_pow_mul_toNat_add (ofBits_lt)]
+
+/-! ### Addition -/
+
+/-- Carry function for bitwise addition. -/
+def bitwise_carry (x y : Nat) : Nat → Bool
+  | 0     => false
+  | i + 1 => (x.testBit i && y.testBit i) || ((x.testBit i ^^ y.testBit i) && bitwise_carry x y i)
+
+/-- Bitwise addition implemented via a ripple carry adder. -/
+@[simp] def bitwise_add (x y i : Nat) :=
```

Comment:
Are you sure you want `@[simp]` here? Usually it is a bad idea to just let definitions fall apart anytime someone mentions `simp` nearby!

Reply:
I see, in every use case when I want to `unfold` the definition I want to go directly to unfolding `ofBits` too (i.e. the intermediate definition isn't useful). Perhaps, there's another way to automatically unfold `ofBits` too whenever I want to unfold without doing `simp only [bitadd, ofBits]`?